### DISCUSSION TITLE
Look for the *.exe.d.json before `lib` when installing app

### DISCUSSION THREAD

**Discussion by iseneirik** [datalust/seq-tickets#1786] [Ideas]
Hi,

After a few days of trying and failing I finally stumbled over the correct way to create an App in dotnet without using the `Seq.Apps`-package, and I have some feedback that might be helpful for future development.

Can the documentation for [Writing Seq Apps in Other Languages](https://docs.datalust.co/docs/seq-apps-in-other-languages) include a table of properties for the "app definition file" along with what the file should be called (found this through examples)

When installing an app from a NuGet, it seems like the "app definition file" is ignored in favor of an existing `lib` folder. So with a working `*.exe`-file with an accompanying `*.exe.d.json`-file, these are simply ignored if the NuGet also includes a `lib` folder with the error: ***No type marked with `[SeqApp]` could be found***. So, could the existence of a `*.exe.d.json`-file be checked first so that the remaining NuGet-contents don't matter?

In general it would be nice with a simple description of how the resulting NuGet for an app should look like:
- What are the minimum required files
- Where should the `*.exe`-file be (root? can it be elsewhere?)
- Where should the "app definition file" be along with what it can be called

**Comment by nblumhardt** [comment]
That's really interesting to know - thanks for sharing your experiences!

The minimum is the `*.d.json` - no other files are required.

The current algorithm does look for `*.d.json` before trying to find/load .NET assemblies. The `*.d.json` file must be in the root of the NUPKG, however - it won't be found in `lib/` or elsewhere.

Does this line up with the behavior you observed? I'll update the linked doc page if so :+1:

Best regards,
Nick

**Reply by iseneirik** [reply]
Hi,

Thanks for the reply 🙂
I tried to create a console application in *net7*, and packaged it in a NUPKG with `*.d.json` located in the root. The first attempt didn't work as the NUPGK had a `lib/`-folder, but when I repackaged it with the only difference being no `lib/`-folder it worked (ran into other issues, but now the app is installable at least). So I guess the question is why does the ***"No type marked with `[SeqApp]` could be found"*** show up? I assume this shows up when the installer expects an app created using the `Seq.Apps`-package? Are there assumptions made when .NET assemblies are found?

**Reply by nblumhardt** [reply]
Ah - okay, that makes sense now! Thanks @iseneirik. Yes, the probe for a `lib/` folder does occur before the attempt to load a definition. I think this is actually a subtle bug in Seq, I'll have a think about it and make some updates accordingly.

Thanks again!

**Reply by iseneirik** [reply]
Awesome, thanks for looking into it! 👍 
I solved my needs by forking the existing `Seq.App.Slack` and making some changes, but if creating custom apps in `net7` becomes easier as a result of this I'll definitely give it a try 🙂

Love using Seq, by the way! Keep up the good work 👊
